File: src/main/java/me/itzisonn_/meazy/parser/Parser.java
package me.itzisonn_.meazy.parser;

import lombok.Getter;
import me.itzisonn_.meazy.lang.text.Text;
import me.itzisonn_.meazy.lexer.Token;
import me.itzisonn_.meazy.lexer.TokenType;
import me.itzisonn_.meazy.lexer.TokenTypeSet;
import me.itzisonn_.meazy.lexer.TokenTypes;
import me.itzisonn_.meazy.parser.ast.Statement;
import me.itzisonn_.meazy.registry.Registries;
import me.itzisonn_.registry.RegistryEntry;
import me.itzisonn_.registry.RegistryIdentifier;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * Is used to store and parse tokens
 * @see Registries#PARSING_FUNCTIONS
 */
@NullMarked
public class Parser {
    private final ParsingContext context;
    private final List<Token> tokens;

    /**
     * Position of current element in {@link Parser#tokens}
     */
    @Getter
    private int pos = 0;

    /**
     *
     * @param context Parsing context
     * @param tokens List of tokens
     */
    public Parser(ParsingContext context, List<Token> tokens) {
        this.context = context;
        this.tokens = tokens;
    }



    /**
     * @return Copy of tokens list
     */
    public List<Token> getTokens() {
        return new ArrayList<>(tokens);
    }

    /**
     * @return Token at {@link Parser#pos} in {@link Parser#tokens}
     */
    public Token getCurrent() {
        return tokens.get(pos);
    }



    /**
     * Increments position by 1
     */
    public void next() {
        pos++;
    }

    /**
     * Increments position by 1
     *
     * @param tokenType Required TokenType
     * @param text Exception's text=
     *
     * @throws UnexpectedTokenException If token's type doesn't match required
     */
    public void next(TokenType tokenType, Text text) throws UnexpectedTokenException {
        if (!getCurrent().getType().equals(tokenType)) {
            throw new UnexpectedTokenException(getCurrent().getLine(), text);
        }

        next();
    }

    /**
     * Increments position by 1
     *
     * @param tokenTypeSet Required TokenTypeSet
     * @param text Exception's text
     *
     * @throws UnexpectedTokenException If tokenTypeSet doesn't contain current token's type
     */
    public void next(TokenTypeSet tokenTypeSet, Text text) throws UnexpectedTokenException {
        if (!tokenTypeSet.getTokenTypes().contains(getCurrent().getType())) {
            throw new UnexpectedTokenException(getCurrent().getLine(), text);
        }

        next();
    }



    /**
     * Returns token at current position and increments position by 1
     * @return Token at {@link Parser#pos} in {@link Parser#tokens}
     */
    public Token getCurrentAndNext() {
        Token token = getCurrent();
        pos++;
        return token;
    }

    /**
     * Returns token at current position increments position by 1
     *
     * @param tokenType Required TokenType
     * @param text Exception's text
     * @return Token at {@link Parser#pos} in {@link Parser#tokens}
     *
     * @throws UnexpectedTokenException If token's type doesn't match required
     */
    public Token getCurrentAndNext(TokenType tokenType, Text text) throws UnexpectedTokenException {
        if (!getCurrent().getType().equals(tokenType)) {
            throw new UnexpectedTokenException(getCurrent().getLine(), text);
        }

        return getCurrentAndNext();
    }

    /**
     * Returns token at current position and increments position by 1
     *
     * @param tokenTypeSet Required TokenTypeSet
     * @param text Exception's text
     * @return Token at {@link Parser#pos} in {@link Parser#tokens}
     *
     * @throws UnexpectedTokenException If tokenTypeSet doesn't contain current token's type
     */
    public Token getCurrentAndNext(TokenTypeSet tokenTypeSet, Text text) throws UnexpectedTokenException {
        if (!tokenTypeSet.getTokenTypes().contains(getCurrent().getType())) {
            throw new UnexpectedTokenException(getCurrent().getLine(), text);
        }

        return getCurrentAndNext();
    }

    /**
     * Skips all {@link TokenTypes#NEW_LINE()} tokens
     */
    public void moveOverOptionalNewLines() {
        while (getCurrent().getType().equals(TokenTypes.NEW_LINE())) pos++;
    }

    /**
     * Checks current line for presence of token with given tokenType
     *
     * @param tokenType Required TokenType
     * @return Whether current line has token with given tokenType
     */
    public boolean currentLineHasToken(TokenType tokenType) {
        for (int i = pos; i < tokens.size(); i++) {
            TokenType current = tokens.get(i).getType();
            if (current.equals(TokenTypes.NEW_LINE())) return false;
            if (current.equals(tokenType)) return true;
        }

        return false;
    }

    /**
     * Checks current line for presence of token with type inside given tokenTypeSet
     *
     * @param tokenTypeSet Required TokenTypeSet
     * @return Whether current line has token with type inside given tokenTypeSet
     */
    public boolean currentLineHasToken(TokenTypeSet tokenTypeSet) {
        for (int i = pos; i < tokens.size(); i++) {
            TokenType current = tokens.get(i).getType();
            if (current.equals(TokenTypes.NEW_LINE())) return false;
            if (tokenTypeSet.contains(current)) return true;
        }

        return false;
    }



    /**
     * Executes ParsingFunction with given id
     *
     * @param id Id of ParsingFunction
     * @param extra Extra info
     * @return Parsed statement
     *
     * @throws IllegalArgumentException When can't find ParsingFunction with given id
     */
    public Statement parse(RegistryIdentifier id, @Nullable Object... extra) throws IllegalArgumentException {
        ParsingFunction<? extends Statement> parsingFunction = getParsingFunctionOrNull(id);
        if (parsingFunction == null) throw new IllegalArgumentException("Can't find ParsingFunction with id " + id);
        return parsingFunction.parse(context, extra);
    }

    /**
     * Executes ParsingFunction with given id
     *
     * @param id Id of ParsingFunction
     * @param cls Required returned statement's class
     * @param extra Extra info
     * @param <T> Returned statement's type
     * @return Parsed statement
     *
     * @throws IllegalArgumentException When can't find ParsingFunction with given id
     *                                  or return type of ParsingFunction doesn't match requested
     */
    @SuppressWarnings("unchecked")
    public <T extends Statement> T parse(RegistryIdentifier id, Class<T> cls, @Nullable Object... extra) throws IllegalArgumentException {
        Statement statement = parse(id, extra);

        if (!cls.isInstance(statement)) {
            throw new IllegalArgumentException("Return type of ParsingFunction with id " + id + " doesn't match requested (" + cls.getName() + ")");
        }

        return (T) statement;
    }

    /**
     * Executes ParsingFunction after ParsingFunction with given id
     *
     * @param id Id of ParsingFunction
     * @param extra Extra info
     * @return Parsed statement
     *
     * @throws IllegalArgumentException When can't find ParsingFunction with given id
     */
    public Statement parseAfter(RegistryIdentifier id, @Nullable Object... extra) throws IllegalArgumentException {
        ParsingFunction<? extends Statement> parsingFunction = getParsingFunctionAfterOrNull(id);
        if (parsingFunction == null) throw new IllegalArgumentException("Can't find ParsingFunction with id " + id);

        return parsingFunction.parse(context, extra);
    }

    /**
     * Executes ParsingFunction after ParsingFunction with given id
     *
     * @param id Id of ParsingFunction
     * @param cls Required returned statement's class
     * @param extra Extra info
     * @param <T> Returned statement's type
     * @return Parsed statement
     *
     * @throws IllegalArgumentException When can't find ParsingFunction with given id
     *                                  or return type of ParsingFunction doesn't match requested
     */
    @SuppressWarnings("unchecked")
    public <T extends Statement> T parseAfter(RegistryIdentifier id, Class<T> cls, @Nullable Object... extra) throws IllegalArgumentException {
        Statement statement = parseAfter(id, extra);

        if (!cls.isInstance(statement)) {
            throw new IllegalArgumentException("Return type of ParsingFunction with id " + id + " doesn't match requested (" + cls.getName() + ")");
        }

        return (T) statement;
    }



    /**
     * Finds ParsingFunction with given id
     *
     * @param id Id
     * @return ParsingFunction with given id or null
     */
    @Nullable
    private ParsingFunction<? extends Statement> getParsingFunctionOrNull(RegistryIdentifier id) {
        RegistryEntry<ParsingFunction<? extends Statement>> entry = Registries.PARSING_FUNCTIONS.getEntry(id);
        if (entry == null) return null;
        return entry.getValue();
    }

    /**
     * Finds ParsingFunction after ParsingFunction with given id
     *
     * @param id Id
     * @return ParsingFunction after ParsingFunction with given id or null
     */
    @Nullable
    private ParsingFunction<? extends Statement> getParsingFunctionAfterOrNull(RegistryIdentifier id) {
        RegistryEntry<ParsingFunction<? extends Statement>> entry = Registries.PARSING_FUNCTIONS.getEntryAfter(id);
        if (entry == null) return null;

        return entry.getValue();
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/parser/Parser.kt
package me.itzisonn_.meazy.parser

import me.itzisonn_.meazy.lexer.Token
import me.itzisonn_.meazy.lexer.TokenBehaviour
import me.itzisonn_.meazy.lexer.TokenType
import me.itzisonn_.meazy.lexer.TokenTypeSet
import me.itzisonn_.meazy.lexer.TokenTypes
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.parser.pasing_function.ParsingFunction
import me.itzisonn_.meazy.util.text.Text

/**
 * Is used to store and parse tokens
 *
 * @param tokens  List of tokens
 */
class Parser(tokens: List<Token>) {
    private val tokens = tokens.toList()

    /**
     * Position of current element
     */
    var pos = 0
        private set

    /**
     * Amount of tokens
     */
    val size get() = tokens.size

    /**
     * @return Token at position [pos]
     */
    private val current get() = tokens[pos]

    /**
     * @return Token at position [i]
     */
    operator fun get(i: Int) = tokens[i]

    fun isEndOfFile() = current.type == TokenTypes.endOfFile



    /**
     * Returns token at current position and increments position by 1
     * @return Token at [pos] in tokens
     */
    fun consume() = current.also { pos++ }

    /**
     * Returns token at current position increments position by 1
     * 
     * @param tokenType Required TokenType
     * @param text      Exception's text
     * @return Token at [pos] in tokens
     *
     * @throws UnexpectedTokenException If token's type doesn't match required
     */
    fun consume(tokenType: TokenType, text: Text?): Token {
        while (current.type != tokenType) {
            if (current.type.behaviour == TokenBehaviour.IGNORE) pos++
            else throw UnexpectedTokenException(text)
        }

        return consume()
    }

    /**
     * Returns token at current position and increments position by 1
     * 
     * @param tokenTypeSet Required TokenTypeSet
     * @param text         Exception's text
     * @return Token at [pos] in tokens
     *
     * @throws UnexpectedTokenException If tokenTypeSet doesn't contain current token's type
     */
    fun consume(tokenTypeSet: TokenTypeSet, text: Text?): Token {
        while (current.type !in tokenTypeSet.getTokenTypes()) {
            if (current.type.behaviour == TokenBehaviour.IGNORE) pos++
            else throw UnexpectedTokenException(text)
        }

        return consume()
    }

    fun isNext(tokenType: TokenType): Boolean {
        val prevPos = pos

        while (current.type != tokenType) {
            if (current.type.behaviour == TokenBehaviour.IGNORE) pos++
            else {
                pos = prevPos
                return false
            }
        }

        pos = prevPos
        return true
    }

    fun isNext(tokenTypeSet: TokenTypeSet): Boolean {
        val prevPos = pos

        while (current.type !in tokenTypeSet) {
            if (current.type.behaviour == TokenBehaviour.IGNORE) pos++
            else {
                pos = prevPos
                return false
            }
        }

        pos = prevPos
        return true
    }

    /** TODO
     * Returns token at current position increments position by 1
     *
     * @param tokenType Required TokenType
     * @param text      Exception's text
     * @return Token at [pos] in tokens
     *
     * @throws UnexpectedTokenException If token's type doesn't match required
     */
    fun find(tokenType: TokenType, text: Text?): Token {
        while (current.type != tokenType) {
            if (current.type.behaviour == TokenBehaviour.IGNORE) pos++
            else throw UnexpectedTokenException(text)
        }

        return current
    }

    fun UnexpectedTokenException(text: Text?) = UnexpectedTokenException(current, text)
    fun InvalidStatementException(text: Text) = InvalidStatementException(current.line, text)
    fun InvalidSyntaxException(text: Text) = InvalidSyntaxException(current.line, text)



    /**
     * Executes given ParsingFunction
     *
     * @param parsingFunction ParsingFunction to execute
     * @param extra           Extra info
     * @param T               Returned program unit's type
     * @return Parsed program unit
     */
    fun <T : ProgramUnit> parse(parsingFunction: ParsingFunction<T>, vararg extra: Any?): T {
        return parsingFunction.run { this@Parser.parse(*extra) }
    }

    /**
     * Executes given ParsingFunction, and if it fails, returns to position before parsing
     *
     * @param parsingFunction ParsingFunction to execute
     * @param extra           Extra info
     * @param T               Returned program unit's type
     * @return Parsed program unit
     */
    fun <T : ProgramUnit> tryParse(parsingFunction: ParsingFunction<T>, vararg extra: Any?): T? {
        val prevPos = pos

        try {
            return parse(parsingFunction, *extra)
        }
        catch (_: UnexpectedTokenException) {
            pos = prevPos
            return null
        }
    }
}